File: core/src/main/java/com/egoal/darkestpixeldungeon/actors/hero/HeroClass.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015  Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2016 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.egoal.darkestpixeldungeon.actors.hero;

import com.egoal.darkestpixeldungeon.DarkestPixelDungeon;
import com.egoal.darkestpixeldungeon.actors.Damage;
import com.egoal.darkestpixeldungeon.items.ArmorKit;
import com.egoal.darkestpixeldungeon.items.CrownOfDwarf;
import com.egoal.darkestpixeldungeon.items.DemonicSkull;
import com.egoal.darkestpixeldungeon.items.DewVial;
import com.egoal.darkestpixeldungeon.items.Gold;
import com.egoal.darkestpixeldungeon.items.armor.MailArmor;
import com.egoal.darkestpixeldungeon.items.artifacts.Astrolabe;
import com.egoal.darkestpixeldungeon.items.TomeOfMastery;
import com.egoal.darkestpixeldungeon.items.UnholyBlood;
import com.egoal.darkestpixeldungeon.items.armor.PlateArmor;
import com.egoal.darkestpixeldungeon.items.ExtractionFlask;
import com.egoal.darkestpixeldungeon.items.artifacts.GoldPlatedStatue;
import com.egoal.darkestpixeldungeon.items.artifacts.HandOfTheElder;
import com.egoal.darkestpixeldungeon.items.artifacts.MasterThievesArmband;
import com.egoal.darkestpixeldungeon.items.artifacts.UrnOfShadow;
import com.egoal.darkestpixeldungeon.items.bags.SeedPouch;
import com.egoal.darkestpixeldungeon.items.books.Book;
import com.egoal.darkestpixeldungeon.items.food.Food;
import com.egoal.darkestpixeldungeon.items.Torch;
import com.egoal.darkestpixeldungeon.Assets;
import com.egoal.darkestpixeldungeon.Badges;
import com.egoal.darkestpixeldungeon.Challenges;
import com.egoal.darkestpixeldungeon.Dungeon;
import com.egoal.darkestpixeldungeon.items.BrokenSeal;
import com.egoal.darkestpixeldungeon.items.armor.ClothArmor;
import com.egoal.darkestpixeldungeon.items.artifacts.CloakOfShadows;
import com.egoal.darkestpixeldungeon.items.food.Wine;
import com.egoal.darkestpixeldungeon.items.potions.*;
import com.egoal.darkestpixeldungeon.items.rings.RingOfAccuracy;
import com.egoal.darkestpixeldungeon.items.rings.RingOfEvasion;
import com.egoal.darkestpixeldungeon.items.rings.RingOfForce;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfCurse;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfLight;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfMagicMapping;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfPsionicBlast;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfTeleportation;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfTerror;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfUpgrade;
import com.egoal.darkestpixeldungeon.items.wands.WandOfBlastWave;
import com.egoal.darkestpixeldungeon.items.wands.WandOfCorruption;
import com.egoal.darkestpixeldungeon.items.wands.WandOfDisintegration;
import com.egoal.darkestpixeldungeon.items.wands.WandOfMagicMissile;
import com.egoal.darkestpixeldungeon.items.wands.WandOfPrismaticLight;
import com.egoal.darkestpixeldungeon.items.weapon.enchantments.Blazing;
import com.egoal.darkestpixeldungeon.items.weapon.melee.AssassinsBlade;
import com.egoal.darkestpixeldungeon.items.weapon.melee.BattleGloves;
import com.egoal.darkestpixeldungeon.items.weapon.melee.CrystalsSwords;
import com.egoal.darkestpixeldungeon.items.weapon.melee.Dagger;
import com.egoal.darkestpixeldungeon.items.weapon.melee.Knuckles;
import com.egoal.darkestpixeldungeon.items.weapon.melee.MagesStaff;
import com.egoal.darkestpixeldungeon.items.weapon.melee.SorceressWand;
import com.egoal.darkestpixeldungeon.items.weapon.melee.Sword;
import com.egoal.darkestpixeldungeon.items.weapon.melee.WornShortsword;
import com.egoal.darkestpixeldungeon.items.weapon.missiles.Boomerang;
import com.egoal.darkestpixeldungeon.items.weapon.missiles.Dart;
import com.egoal.darkestpixeldungeon.messages.Messages;
import com.egoal.darkestpixeldungeon.plants.Blindweed;
import com.egoal.darkestpixeldungeon.plants.Earthroot;
import com.egoal.darkestpixeldungeon.plants.Firebloom;
import com.egoal.darkestpixeldungeon.plants.Sorrowmoss;
import com.watabou.utils.Bundle;

public enum HeroClass {

  WARRIOR("warrior"),
  MAGE("mage"),
  ROGUE("rogue"),
  HUNTRESS("huntress"),

  SORCERESS("sorceress");

  private String title;

  HeroClass(String title) {
    this.title = title;
  }

  public void initHero(Hero hero) {

    hero.heroClass = this;

    initCommon(hero);

    switch (this) {
      case WARRIOR:
        initWarrior(hero);
        break;

      case MAGE:
        initMage(hero);
        break;

      case ROGUE:
        initRogue(hero);
        break;

      case HUNTRESS:
        initHuntress(hero);
        break;

      //
      case SORCERESS:
        initSorceress(hero);
        break;
    }

    initPerks(hero);
    hero.updateAwareness();
  }

  private static void initCommon(Hero hero) {
    if (!Dungeon.isChallenged(Challenges.NO_ARMOR))
      (hero.belongings.armor = new ClothArmor()).identify();

    if (!Dungeon.isChallenged(Challenges.NO_FOOD))
      new Food().identify().collect();

    if (!Dungeon.isChallenged(Challenges.DARKNESS))
      new Torch().identify().collect();

    if (DarkestPixelDungeon.debug()) {
      initDebug(hero);
    }
    
  }

  private static void initDebug(Hero hero) {
    hero.HP = 1000;
    hero.HT = 1000;
    hero.STR  = 18;
    hero.lvl  = 20;
    
    new ScrollOfMagicMapping().quantity(9).identify().collect();
    new ScrollOfLight().quantity(3).identify().collect();
    new ScrollOfCurse().quantity(3).identify().collect();
    
    new PotionOfHealing().quantity(9).identify().collect();
    new PotionOfMindVision().quantity(9).identify().collect();
    new PotionOfLiquidFlame().quantity(2).identify().collect();
    new PotionOfToxicGas().quantity(3).identify().collect();
    new PotionOfInvisibility().quantity(3).identify().collect();
    
    (new PlateArmor()).upgrade(9).identify().collect();
    (new AssassinsBlade()).upgrade(9).identify().collect();
    new MailArmor().upgrade(1).identify().collect();
    new Sword().upgrade(1).identify().collect();
    
    new Earthroot.Seed().quantity(3).identify().collect();
    
    (new TomeOfMastery()).identify().collect();
    (new ArmorKit()).identify().collect();

    (new SeedPouch()).identify().collect();
    
    new WandOfBlastWave().identify().collect();
    new WandOfCorruption().upgrade(10).identify().collect();
    new WandOfDisintegration().upgrade(2).identify().collect();
    
    // rings
    new HandOfTheElder().identify().collect();
    new RingOfEvasion().identify().collect();
    new RingOfForce().identify().collect();
    new RingOfForce().identify().collect();
    new RingOfAccuracy().identify().collect();
    
    new MasterThievesArmband().identify().collect();
  }

  public Badges.Badge masteryBadge() {
    switch (this) {
      case WARRIOR:
        return Badges.Badge.MASTERY_WARRIOR;
      case MAGE:
        return Badges.Badge.MASTERY_MAGE;
      case ROGUE:
        return Badges.Badge.MASTERY_ROGUE;
      case HUNTRESS:
        return Badges.Badge.MASTERY_HUNTRESS;
      case SORCERESS:
        return Badges.Badge.MASTERY_SORCERESS;
    }
    return null;
  }

  private static void initWarrior(Hero hero) {
    (hero.belongings.weapon = new WornShortsword()).identify();
    Dart darts = new Dart(8);
    darts.identify().collect();

    new Wine().collect();
    
    if (Badges.isUnlocked(Badges.Badge.TUTORIAL_WARRIOR)) {
      if (!Dungeon.isChallenged(Challenges.NO_ARMOR))
        hero.belongings.armor.affixSeal(new BrokenSeal());
      Dungeon.quickslot.setSlot(0, darts);
    } else {
      if (!Dungeon.isChallenged(Challenges.NO_ARMOR)) {
        BrokenSeal seal = new BrokenSeal();
        seal.collect();
        Dungeon.quickslot.setSlot(0, seal);
      }
      Dungeon.quickslot.setSlot(1, darts);
    }

    // new perk
    hero.heroPerk.add(HeroPerk.Perk.DRUNKARD);

    new PotionOfHealing().setKnown();
    
    // resists
    hero.addResistances(Damage.Element.FIRE, 1.1f);
    hero.addResistances(Damage.Element.LIGHT, .9f);
    hero.addResistances(Damage.Element.SHADOW, .8f, .9f);
  }

  private static void initMage(Hero hero) {
    MagesStaff staff;

    if (Badges.isUnlocked(Badges.Badge.TUTORIAL_MAGE)) {
      staff = new MagesStaff(new WandOfMagicMissile());
    } else {
      staff = new MagesStaff();
      new WandOfMagicMissile().identify().collect();
    }

    (hero.belongings.weapon = staff).identify();
    hero.belongings.weapon.activate(hero);

    Dungeon.quickslot.setSlot(0, staff);

    new ScrollOfUpgrade().setKnown();
    
    hero.addResistances(Damage.Element.FIRE, 1f, 1.2f);
    hero.addResistances(Damage.Element.POISON, .8f);
    hero.addResistances(Damage.Element.LIGHT, 1.1f);
  }

  private static void initRogue(Hero hero) {
    (hero.belongings.weapon = new Dagger()).identify();

    CloakOfShadows cloak = new CloakOfShadows();
    (hero.belongings.misc1 = cloak).identify();
    hero.belongings.misc1.activate(hero);

    Dart darts = new Dart(8);
    darts.identify().collect();

    Dungeon.quickslot.setSlot(0, cloak);
    Dungeon.quickslot.setSlot(1, darts);

    hero.heroPerk.add(HeroPerk.Perk.CRITICAL_STRIKE);
    hero.heroPerk.add(HeroPerk.Perk.KEEN);

    new ScrollOfMagicMapping().setKnown();
    
    hero.addResistances(Damage.Element.POISON, 1.2f);
    hero.addResistances(Damage.Element.ICE, .8f, .9f);
    hero.addResistances(Damage.Element.SHADOW, .9f, 1.1f);
  }

  private static void initHuntress(Hero hero) {

    (hero.belongings.weapon = new Knuckles()).identify();
    Boomerang boomerang = new Boomerang();
    boomerang.identify().collect();

    Dungeon.quickslot.setSlot(0, boomerang);

    hero.heroPerk.add(HeroPerk.Perk.NIGHT_VISION);
    hero.heroPerk.add(HeroPerk.Perk.SHOOTER);

    new PotionOfMindVision().setKnown();
    
    hero.addResistances(Damage.Element.POISON, 1.2f);
    hero.addResistances(Damage.Element.ICE, .9f);
    hero.addResistances(Damage.Element.ACID, .9f);
  }

  // extra classes
  private static void initSorceress(Hero hero) {
    // perks
    hero.heroPerk.add(HeroPerk.Perk.SHREWD);
    hero.heroPerk.add(HeroPerk.Perk.POSITIVE);

    (hero.belongings.weapon = new SorceressWand()).identify();

    ExtractionFlask flask = new ExtractionFlask();
    flask.identify().collect();

    // ranged weapon
    Dart darts = new Dart(6);
    darts.identify().collect();

    Dungeon.quickslot.setSlot(0, flask);
    Dungeon.quickslot.setSlot(1, darts);

    new PotionOfToxicGas().identify().collect();

    // resists and extra resists to poison
    for (int i = 0; i < Damage.Element.ELEMENT_COUNT; ++i) {
      hero.addResistances(1 << i, 1.25f, 1f);
    }
    hero.addResistances(Damage.Element.POISON, 2f);    
  }

  private static void initPerks(Hero hero) {
    if (hero.heroPerk.contain(HeroPerk.Perk.CRITICAL_STRIKE))
      hero.criticalChance_ += 5f / 100f;

  }

  public String title() {
    return Messages.get(HeroClass.class, title);
  }

  public String spritesheet() {

    switch (this) {
      case WARRIOR:
        return Assets.WARRIOR;
      case MAGE:
        return Assets.MAGE;
      case ROGUE:
        return Assets.ROGUE;
      case HUNTRESS:
        return Assets.HUNTRESS;
      case SORCERESS:
        return Assets.DPD_SORCERESS;
    }

    return null;
  }

  public String[] perks() {

    switch (this) {
      case WARRIOR:
        return new String[]{
                Messages.get(HeroClass.class, "warrior_perk1"),
                Messages.get(HeroClass.class, "warrior_perk2"),
                Messages.get(HeroClass.class, "warrior_perk3"),
                Messages.get(HeroClass.class, "warrior_perk4"),
                Messages.get(HeroClass.class, "warrior_perk5"),
        };
      case MAGE:
        return new String[]{
                Messages.get(HeroClass.class, "mage_perk1"),
                Messages.get(HeroClass.class, "mage_perk2"),
                Messages.get(HeroClass.class, "mage_perk3"),
                Messages.get(HeroClass.class, "mage_perk4"),
                Messages.get(HeroClass.class, "mage_perk5"),
        };
      case ROGUE:
        return new String[]{
                Messages.get(HeroClass.class, "rogue_perk1"),
                Messages.get(HeroClass.class, "rogue_perk2"),
                Messages.get(HeroClass.class, "rogue_perk3"),
                Messages.get(HeroClass.class, "rogue_perk4"),
                Messages.get(HeroClass.class, "rogue_perk5"),
                Messages.get(HeroClass.class, "rogue_perk6"),
        };
      case HUNTRESS:
        return new String[]{
                Messages.get(HeroClass.class, "huntress_perk1"),
                Messages.get(HeroClass.class, "huntress_perk2"),
                Messages.get(HeroClass.class, "huntress_perk3"),
                Messages.get(HeroClass.class, "huntress_perk4"),
                Messages.get(HeroClass.class, "huntress_perk5"),
        };
      case SORCERESS:
        return new String[]{
                Messages.get(HeroClass.class, "sorceress_perk1"),
                Messages.get(HeroClass.class, "sorceress_perk2"),
                Messages.get(HeroClass.class, "sorceress_perk3"),
                Messages.get(HeroClass.class, "sorceress_perk4"),
                Messages.get(HeroClass.class, "sorceress_perk5"),
                // Messages.get(HeroClass.class, "sorceress_perk6"),
        };
    }

    return null;
  }

  private static final String CLASS = "class";

  public void storeInBundle(Bundle bundle) {
    bundle.put(CLASS, toString());
  }

  public static HeroClass restoreInBundle(Bundle bundle) {
    String value = bundle.getString(CLASS);
    return value.length() > 0 ? valueOf(value) : ROGUE;
  }
}

File: core/src/main/java/com/egoal/darkestpixeldungeon/actors/hero/HeroClass.kt
package com.egoal.darkestpixeldungeon.actors.hero

import com.egoal.darkestpixeldungeon.*
import com.egoal.darkestpixeldungeon.actors.Damage
import com.egoal.darkestpixeldungeon.actors.buffs.Pressure
import com.egoal.darkestpixeldungeon.actors.hero.perks.*
import com.egoal.darkestpixeldungeon.effects.PerkGain
import com.egoal.darkestpixeldungeon.items.armor.*
import com.egoal.darkestpixeldungeon.items.armor.glyphs.Peaceful
import com.egoal.darkestpixeldungeon.items.artifacts.*
import com.egoal.darkestpixeldungeon.items.bags.SeedPouch
import com.egoal.darkestpixeldungeon.items.bags.SkillTree
import com.egoal.darkestpixeldungeon.items.food.Food
import com.egoal.darkestpixeldungeon.items.food.Wine
import com.egoal.darkestpixeldungeon.items.helmets.CollarOfSlave
import com.egoal.darkestpixeldungeon.items.helmets.Headgear
import com.egoal.darkestpixeldungeon.items.helmets.RangerHat
import com.egoal.darkestpixeldungeon.items.potions.*
import com.egoal.darkestpixeldungeon.items.scrolls.*
import com.egoal.darkestpixeldungeon.items.specials.Astrolabe
import com.egoal.darkestpixeldungeon.items.specials.Penetration
import com.egoal.darkestpixeldungeon.items.specials.Shadowmoon
import com.egoal.darkestpixeldungeon.items.specials.UrnOfShadow
import com.egoal.darkestpixeldungeon.items.unclassified.*
import com.egoal.darkestpixeldungeon.items.wands.*
import com.egoal.darkestpixeldungeon.items.weapon.Weapon
import com.egoal.darkestpixeldungeon.items.weapon.melee.*
import com.egoal.darkestpixeldungeon.items.weapon.missiles.*
import com.egoal.darkestpixeldungeon.messages.M
import com.egoal.darkestpixeldungeon.messages.Messages
import com.egoal.darkestpixeldungeon.plants.CorrodeCyan
import com.egoal.darkestpixeldungeon.utils.GLog
import com.watabou.utils.Bundle
import com.watabou.utils.Random
import kotlin.math.min

enum class HeroClass(private val title: String, vararg subclasses: HeroSubClass) {
    WARRIOR("warrior", HeroSubClass.GLADIATOR, HeroSubClass.BERSERKER, HeroSubClass.KNIGHTT) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_WARRIOR
        override fun spritesheet(): String = Assets.WARRIOR
        override fun perks(): List<String> = (1..3).map { Messages.get(HeroClass::class.java, "warrior_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(Drunkard(), GoodAppetite(), RavenousAppetite())

        override fun initHeroStatus(hero: Hero) {
            super.initHeroStatus(hero)
//            hero.HP += 5 // 25
//            hero.HT += 5
            hero.MSHLD += 3
        }

        override fun onHeroUpgraded(hero: Hero) {
            hero.HP += 1 // extra +1
            hero.HT += 1
        }

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)
            hero.belongings.weapon = WornShortsword().identify() as Weapon

            val darts = Dart(4)
            darts.identify().collect()

            Wine().collect()

            if (Badges.isUnlocked(Badges.Badge.TUTORIAL_WARRIOR)) {
                hero.belongings.armor!!.affixSeal(BrokenSeal())
                Dungeon.quickslot.setSlot(0, darts)
            } else {
                val seal = BrokenSeal()
                seal.collect()
                Dungeon.quickslot.setSlot(0, seal)
                Dungeon.quickslot.setSlot(1, darts)
            }

            PotionOfStrength().setKnown()

            // resists
            hero.addResistances(Damage.Element.FIRE, 0.1f)
            hero.addResistances(Damage.Element.LIGHT, -0.1f)
            hero.addResistances(Damage.Element.SHADOW, -0.2f)
        }
    },

    MAGE("mage", HeroSubClass.BATTLEMAGE, HeroSubClass.WARLOCK, HeroSubClass.ARCHMAGE) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_MAGE
        override fun spritesheet(): String = Assets.MAGE
        override fun perks(): List<String> = (1..3).map { Messages.get(HeroClass::class.java, "mage_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(GoodAppetite(), WandPerception(), PreheatedZap())

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)

//            val staff =
//                    if (Badges.isUnlocked(Badges.Badge.TUTORIAL_MAGE)) MagesStaff(WandOfMagicMissile())
//                    else {
//                        WandOfMagicMissile().identify().collect()
//                        MagesStaff()
//                    }
//
//            hero.belongings.weapon = staff
//            staff.identify()
//            staff.activate(hero)
            hero.belongings.weapon = ParryingDagger().identify() as Weapon

            val staff = WandOfMagicMissile().identify()
            staff.collect()
            Dungeon.quickslot.setSlot(0, staff)

            ScrollOfUpgrade().setKnown()

            hero.addResistances(Damage.Element.FIRE, 0.1f)
            hero.addResistances(Damage.Element.POISON, -0.2f)
            hero.addResistances(Damage.Element.LIGHT, -0.1f)
            hero.addResistances(Damage.Element.SHADOW, 0.1f)
        }

        override fun upgradeHero(hero: Hero) {
            super.upgradeHero(hero)
            if (hero.lvl == 12) {
                val p = WandPerception()
                if (p.isAcquireAllowed(hero)) {
                    hero.heroPerk.add(p)
                    PerkGain.Show(hero, p)

                    GLog.p(M.L(p, "gain_level_2"))
                }
            }
        }
    },

    ROGUE("rogue", HeroSubClass.FREERUNNER, HeroSubClass.ASSASSIN) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_ROGUE
        override fun spritesheet(): String = Assets.ROGUE
        override fun perks(): List<String> = (1..4).map { Messages.get(HeroClass::class.java, "rogue_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(LowWeightDexterous(), Dieting(), ExtraCritProbability(), Keen())

        override fun onHeroUpgraded(hero: Hero) {
            hero.criticalChance += 0.1f / 100f
        }

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)

            hero.belongings.weapon = Dagger().identify() as Weapon

            val cloak = CloakOfShadows()
            hero.belongings.misc1 = cloak
            cloak.identify()
            cloak.activate(hero)

            val darts = Dart(8)
            darts.identify().collect()

            Dungeon.quickslot.setSlot(0, cloak)
            Dungeon.quickslot.setSlot(1, darts)

            ScrollOfMagicMapping().setKnown()

            hero.addResistances(Damage.Element.POISON, 0.2f)
            hero.addResistances(Damage.Element.ICE, -0.2f)
            hero.addResistances(Damage.Element.SHADOW, -0.1f)
        }
    },

    HUNTRESS("huntress", HeroSubClass.SNIPER, HeroSubClass.WARDEN, HeroSubClass.MOONRIDER) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_HUNTRESS
        override fun spritesheet(): String = Assets.HUNTRESS
        override fun perks(): List<String> = (1..4).map { Messages.get(HeroClass::class.java, "huntress_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(NightVision(), Telepath())

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)

            hero.belongings.weapon = Knuckles().identify() as Weapon

//            val rh = RangersHook()
//            hero.belongings.misc1 = rh
//            rh.identify()
//            rh.activate(hero)
//            Dungeon.quickslot.setSlot(0, rh)
;
            val b = Boomerang()
            b.identify().collect()
            Dungeon.quickslot.setSlot(0, b)

            PotionOfMindVision().setKnown()

            hero.addResistances(Damage.Element.POISON, 0.2f)
            hero.addResistances(Damage.Element.ICE, -0.1f)
            hero.addResistances(Damage.Element.FIRE, -0.2f)
        }
    },

    SORCERESS("sorceress", HeroSubClass.STARGAZER, HeroSubClass.WITCH) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_SORCERESS
        override fun spritesheet(): String = Assets.DPD_SORCERESS
        override fun perks(): List<String> = (1..4).map { Messages.get(HeroClass::class.java, "sorceress_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(GoodAppetite(), Discount(), Optimistic())

        override fun onHeroUpgraded(hero: Hero) {
            hero.HT -= 1
            hero.HP -= 1
            hero.regeneration += 0.01f
        }

        override fun initHeroStatus(hero: Hero) {
            super.initHeroStatus(hero)

            hero.regeneration += 0.02f
            hero.magicalResistance = 0.15f
            hero.addResistances(Damage.Element.all(), 0.1f)
            hero.addResistances(Damage.Element.POISON, 0.5f)
        }

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)

            hero.belongings.weapon = SorceressWand().identify() as Weapon

            val flask = ExtractionFlask()
            flask.identify().collect()
            Dungeon.quickslot.setSlot(0, flask)

            val darts = Salt(6)
            darts.identify().collect()
            Dungeon.quickslot.setSlot(1, darts)

            PotionOfToxicGas().identify().collect()
        }
    },

    EXILE("exile", HeroSubClass.LANCER, HeroSubClass.WINEBIBBER) {
        override fun masteryBadge(): Badges.Badge = Badges.Badge.MASTERY_EXILE
        override fun spritesheet(): String = Assets.EXILE
        override fun perks(): List<String> = (1..3).map { Messages.get(HeroClass::class.java, "exile_perk$it") }
        override fun initialPerks(): List<Perk> = listOf(LowHealthRegeneration(), Discount().apply { level = -1 }, PolearmMaster())

        override fun initHeroStatus(hero: Hero) {
            super.initHeroStatus(hero)

            hero.magicalResistance = 0.04f
            hero.addResistances(Damage.Element.ICE, 0.25f)
            hero.addResistances(Damage.Element.FIRE, -0.2f)
            hero.addResistances(Damage.Element.SHADOW, 0.2f)
            hero.addResistances(Damage.Element.LIGHT, -0.1f)
        }

        override fun initHeroClass(hero: Hero) {
            super.initHeroClass(hero)

            hero.belongings.weapon = ShortSpear().identify() as Weapon

            val gourd = Gourd()
            gourd.identify().collect()
            Dungeon.quickslot.setSlot(0, gourd)

            val fc = FlyCutter(3)
            fc.identify().collect()
            Dungeon.quickslot.setSlot(1, fc)

            ScrollOfRage().identify()
        }
    }
    ;

    fun initHero(hero: Hero) {
        hero.heroClass = this

        initHeroStatus(hero)
        initHeroClass(hero)

        if (DarkestPixelDungeon.debug()) initDebug(hero)
    }

    fun title(): String = Messages.get(HeroClass::class.java, title)

    fun desc(): String = M.L(HeroClass::class.java, "${title}_desc")

    val subClasses = subclasses.toList()

    abstract fun masteryBadge(): Badges.Badge
    abstract fun spritesheet(): String
    abstract fun perks(): List<String>
    abstract fun initialPerks(): List<Perk>

    fun storeInBundle(bundle: Bundle) {
        bundle.put(CLASS, toString())
    }

    // common
    protected open fun initHeroStatus(hero: Hero) {
        hero.atkSkill = 10f
        hero.defSkill = 5f
        hero.magicalResistance = 0.08f
    }

    protected open fun initHeroClass(hero: Hero) {
        hero.belongings.armor = ClothArmor().identify() as Armor

        Food().collect()
        Dungeon.torch += 1f

        ScrollOfIdentify().setKnown()

        SeedPouch().identify().collect()
        Dungeon.limitedDrops.seedBag.drop()

        for (p in initialPerks()) hero.heroPerk.add(p)

        SkillTree().identify().collect()

//        ChaliceOfBlood().collect()


//        initDebug(hero)
    }

    // called when hero level up
    open fun upgradeHero(hero: Hero) {
        val ht = hero.HT
        val hp = hero.HP

        hero.apply {
            lvl++
            HT += 5
            HP += 5

            atkSkill++
            defSkill += 0.75f

            criticalChance += 0.4f / 100f
            regeneration += 0.03f

            recoverSanity(min(Random.NormalIntRange(1, lvl * 3 / 4).toFloat(),
                    buff(Pressure::class.java)!!.pressure * 0.3f))
        }

        hero.heroPerk.get(StrongConstitution::class.java)?.upgradeHero(hero)
        hero.heroPerk.get(ExtraDexterousGrowth::class.java)?.upgradeHero(hero)

        onHeroUpgraded(hero)

        if (hero.challenge == Challenge.Faith || hero.challenge == Challenge.Immortality) {
            hero.HP = hp
            hero.HT = ht
        }
        if (hero.challenge == Challenge.Faith)
            Ankh().apply { isBlessed = true }.collect()
    }

    protected open fun onHeroUpgraded(hero: Hero) {}

    private fun initDebug(hero: Hero) {
        // for (i in 1..20) upgradeHero(hero)
        hero.atkSkill = 30f
        hero.defSkill = 30f
        hero.STR = 20
        hero.HT = 1000
        hero.HP = hero.HT

        hero.heroPerk.add(IntendedTransportation())

        Dungeon.quickslot.setSlot(6, FlyCutter().apply {
            quantity(99).identify().collect()
        })
        Dungeon.quickslot.setSlot(5, ScrollOfMagicMapping().apply {
            quantity(99).identify().collect()
        })
        Dungeon.quickslot.setSlot(4, ScrollOfTeleportation().apply {
            quantity(99).identify().collect()
        })
        Dungeon.quickslot.setSlot(3, PotionOfMindVision().apply {
            quantity(99).identify().collect()
        })

        PotionOfHealing().quantity(99).identify().collect()
        PotionOfLiquidFlame().quantity(99).identify().collect()
        PotionOfExperience().quantity(99).identify().collect()

        PlateArmor().identify().upgrade(6).collect()
        Claymore().identify().upgrade(6).collect()

        Headgear().collect()

        Amulet().collect()
    }

    companion object {
        private const val CLASS = "class"

        fun RestoreFromBundle(bundle: Bundle): HeroClass {
            val value = bundle.getString(CLASS)
            return if (value.isNotEmpty()) valueOf(value) else ROGUE
        }
    }

}